### PR TITLE
fix: no more dataurls & localhost links

### LINKED ISSUE #899: Wrong image links allowed
I'm not sure if we can call it an issue but:
- We allow base64 contents as screenshot URL, making shipyard crash for someone
- We also allow 32K image, it might make someone else crash, but didn't test it out

### LINKED ISSUE #901: Users allowed to submit a localhost URL as a project demo
The user gave a localhost url as a demo: https://hackclub.slack.com/archives/C07PZNMBPBN/p1733103195394949

I also saw the same project

### PR DESCRIPTION
**Pull request by NeonGamerBot-QK**
Fixes #899 #901 
Closes #899 
Closes #901 
this also makes it so only `http(s)` urls work

### CODE REVIEW

**Comment by malted** [comment]
Wait up!

File: src/app/harbor/shipyard/new-ship-form.tsx [hunk 1]
```diff
@@ -157,6 +157,28 @@ export default function NewShipForm({
       return
     }
 
+    if (!screenshotUrl.startsWith('http')) {
```
Comment (malted) [inline src/app/harbor/shipyard/new-ship-form.tsx:160]: Won't this fire for `https` links too? `https://` starts with `http`!

Comment (NeonGamerBot-QK) [reply src/app/harbor/shipyard/new-ship-form.tsx:160]: should both http and https links be allowed or just http?

File: src/middleware.ts [hunk 2]
```diff
@@ -22,7 +22,13 @@ async function loadShipsCookie(
...
       response.cookies.set({
         name: 'ships',
-        value: JSON.stringify(ships),
+        value: JSON.stringify(
+          ships.map((s) => {
+            if (s.screenshotUrl.startsWith('data:'))
+              s.screenshotUrl = `replace_me_with_a_non_dataencoded_url_plz`
+            return s
+          }),
+        ),
```
Comment (malted) [inline src/middleware.ts:31]: Is this to fix data that is currently incorrect in the wild?

Comment (NeonGamerBot-QK) [reply src/middleware.ts:31]: yes! currently it soft locks users (aka vercel 500) which means they cant ship projects. ofc i can remove this and you guys will have to unlock them manually ;-;

File: src/app/harbor/shipyard/new-ship-form.tsx (hunk 1 above)
Comment (malted) [reply src/app/harbor/shipyard/new-ship-form.tsx:160]: Just https.

File: src/middleware.ts (hunk 2 above)
Comment (malted) [reply src/middleware.ts:31]: Ok amazing - tysm for this!

File: src/app/harbor/shipyard/new-ship-form.tsx [hunk 3]
```diff
@@ -157,6 +157,28 @@ export default function NewShipForm({
...
+      deploymentUrl.includes('127.0.0.1')
+    ) {
+      toast({
+        title: "That's not a demo link!",
+        description:
+          'Please make sure your link isnt a local link.. Please submit a deployed link instead!',
+      })
+      setStaging(false)
+      return
+    }
```
Comment (malted) [inline src/app/harbor/shipyard/new-ship-form.tsx:180]:
```suggestion
```

Prefer https://github.com/hackclub/high-seas/pull/902/files#diff-67f6ef525eb8d069479999e787507ff34a7193558c62e2f3d6fc7419e740a51fR103-R107

**Comment by maxwofford** [comment]
@malted I'll leave this for you to merge then!

**Pull request merged**